### PR TITLE
[Merged by Bors] - Fix unsoundness with `Or`/`AnyOf`/`Option` component access'

### LINKED ISSUE #4657: Using `Without` in `Or` filters is unsound
## Bevy version

Current main (but should also happen with 0.7 release)

## What you did

```rust
use bevy::prelude::*;

#[derive(Component, Debug)]
struct Foo(u32);

#[derive(Component, Debug)]
struct Bar(u32);

fn setup(mut commands: Commands) {
    commands.spawn().insert(Foo(1));
}

fn test(
    mut query_1: Query<&mut Foo, Or<(Without<Foo>, Without<Bar>)>>,
    mut query_2: Query<&mut Foo, Without<Bar>>,
) {
    let mut x = query_1.single_mut();
    let mut y = query_2.single_mut();
    x.0 = 1;
    println!("{:?} {:?}", *x, *y);
    y.0 = 2;
    println!("{:?} {:?}", *x, *y);
}

fn main() {
    App::new()
        .add_plugins(MinimalPlugins)
        .add_startup_system(setup)
        .add_system(test)
        .run();
}
```

## What you expected to happen

`test` should be disallowed as it has conflicting queries.

## What actually happened

The code runs successfully and prints `Foo(1) Foo(1)` and `Foo(2) Foo(2)` repeatedly.

## Additional information

Current `Or<Without<A>, Without<B>>` get recorded as `access.add_without(A); access.add_without(B);`, which is the cause of problems here. It is probably too difficult to express the "or" condition precisely (filters can encode arbitrary boolean expressions), and a good solution is to handle such a case as if no filters are given.

### PR DESCRIPTION
**Pull request by BoxyUwU**
# Objective

Fixes #4657

Example code that wasnt panic'ing before this PR (and so was unsound):
```rust
    #[test]
    #[should_panic = "error[B0001]"]
    fn option_has_no_filter_with() {
        fn sys(_1: Query<(Option<&A>, &mut B)>, _2: Query<&mut B, Without<A>>) {}
        let mut world = World::default();
        run_system(&mut world, sys);
    }

    #[test]
    #[should_panic = "error[B0001]"]
    fn any_of_has_no_filter_with() {
        fn sys(_1: Query<(AnyOf<(&A, ())>, &mut B)>, _2: Query<&mut B, Without<A>>) {}
        let mut world = World::default();
        run_system(&mut world, sys);
    }

    #[test]
    #[should_panic = "error[B0001]"]
    fn or_has_no_filter_with() {
        fn sys(_1: Query<&mut B, Or<(With<A>, With<B>)>>, _2: Query<&mut B, Without<A>>) {}
        let mut world = World::default();
        run_system(&mut world, sys);
    }
```
## Solution

- Only add the intersection of `with`/`without` accesses of all the elements in `Or/AnyOf` to the world query's `FilteredAccess<ComponentId>` instead of the union.
- `Option`'s fix can be thought of the same way since its basically `AnyOf<T, ()>` but its impl is just simpler as `()` has no `with`/`without` accesses
---

## Changelog

- `Or`/`AnyOf`/`Option` will now report more query conflicts in order to fix unsoundness

## Migration Guide

- If you are now getting query conflicts from `Or`/`AnyOf`/`Option` rip to you and ur welcome for it now being caught

### CODE REVIEW

File: crates/bevy_ecs/src/query/filter.rs
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/query/filter.rs:458]: Is there a good way to deduplicate this code? I'd like to try and avoid having to manually maintain two copies of soundness critical code.

Comment (BoxyUwU) [reply crates/bevy_ecs/src/query/filter.rs:458]: not in this PR pls, in general tho `AnyOf` and `Or` are basically duplicate code :thinking: it'd be nice to deduplicate them (I think with some of the ptrification refactorings to remove `FilterFetch` we actually can get quite close to this)

**Comment by Nilirad** [comment]
Please I want that migration guide on the next release.

File: crates/bevy_ecs/src/query/fetch.rs
Comment (cart) [inline crates/bevy_ecs/src/query/fetch.rs:1094]: Ignoring filters here is safe because we're "increasing what we don't allow", but I think its important to call out that this makes us stricter than we theoretically need to be:

```rust
#[test]
fn this_is_now_too_strict_and_fails_when_it_doesnt_need_to() {
    fn sys(
        x: Query<(&mut A, Option<(&mut C, With<B>)>)>,
        y: Query<&mut C, (Without<B>, With<A>)>,
    ) {
    }
    let mut world = World::default();
    run_system(&mut world, sys);
}
```

Query `x` writes A (all archetypes) and only writes Cs on archetypes that have A and B. Query `y` only writes Cs on archetypes without Bs and with As. This means the C writes are disjoint. But because we're ignoring filter information we don't allow this.

I do think this particular case is pretty niche and solving soundness takes priority. But if we agree that this is "overly strict" its worth recording as a comment here. I haven't yet thought about solutions to this problem, but I suspect that our "flattened filtered accesses" might prevent this level of granularity? Idk if moving to a more complicated model is worth the complexity / computation costs. Definitely don't need to solve this problem in this pr (at least in the context of Option ... about to start looking at AnyOf/Or).

This also kind of breaks my brain, so lemme know if i miscalculated :)

Comment (BoxyUwU) [reply crates/bevy_ecs/src/query/fetch.rs:1094]: Yea you're right that `Option<(&mut C, With<B>)>` isnt handled as efficiently as possible (slash the current impl is overly restrictive), I'm not sure exactly how we'd get this to get work.

Comment (alice-i-cecile) [reply crates/bevy_ecs/src/query/fetch.rs:1094]: Agreed on the analysis here. I have ideas for a more grounded refactor of this space; I'll see if I can tackle it there.

Definitely don't block on this.

File: crates/bevy_ecs/src/query/fetch.rs
Comment (cart) [inline crates/bevy_ecs/src/query/fetch.rs:1688]: Up until this point we've been recording "filtered access" reads as a combination of "with" and "reads_and_writes" entries for a given component id (the add_read api adds both). This effectively decouples the two and pretends something can "read" an id while simultaneously not having a corresponding "with" for the id.

Given that this whole system was built with that behavior in mind, changing the assumption feels risky. I haven't fully explored the thought yet, just curious if you've run through all of the scenarios here.

Comment (cart) [reply crates/bevy_ecs/src/query/fetch.rs:1688]: I'm pretty sure the existence of with / without is currently only used to allow things to not conflict that normally would. If that is completely true, then the only risk is making more things conflict unnecessarily (so this isn't a soundness risk). Playing around with scenarios now to convince myself this is ok.

Comment (cart) [reply crates/bevy_ecs/src/query/fetch.rs:1688]: Ok I'm reasonably convinced this won't break anything. Although this is "hackey" enough that I think its worth expanding on how and why it violates the reads/with coupling. We should probably also add a note to filtered access that "read" should not be assumed to imply "with" in some cases.

**Review by cart — APPROVED** [review]
I'm on board for this fix. As mentioned in some comments, I think we should add some docs to capture some of the weirder stuff going on here. I'll be gone for a wedding starting tomorrow, so maintainers should feel free to merge this when those docs are added.

**Review by alice-i-cecile — APPROVED** [review]

**Pull request closed without merging**
